v2: Add host item executor and block iterator

rvolosatovs opened the pull request:
Host block executor for enarx/sallyport#17 

Blocked by ~#26~ enarx/sallyport#28 

- Added a custom `Iterator` trait, which supports consuming, tail-recursive implementations
- Added example `exit` call handler.
- Updated the toolchain used by CI to allow using const generics.

Commit:
ci: Update toolchain 1.50.0 -> 1.57.0

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2/item: Implement `TryFrom<usize>` for `Kind`

Co-authored-by: Harald Hoyer <harald@profian.com>
Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2/item: Implement `TryFrom<[usize; 2]>` for `Header`

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2/item: Implement Iterator over `Item` instances

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2: Add optional `asm` feature

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2/host: Add module

Signed-off-by: Roman Volosatovs <roman@profian.com>

rvolosatovs commented:
Including #28 here, perhaps it makes more sense to also see how it's used and why it's required.

Commit:
v2: Mark `item` module public outside crate boundary

Signed-off-by: Roman Volosatovs <roman@profian.com>

haraldh reviewed (commented):
LGTM... waiting on @npmccallum

haraldh commented:
Even passes miri:
```
❯ cargo +nightly miri test
[…]
 Running unittests (target/miri/x86_64-unknown-linux-gnu/debug/deps/sallyport-bed894b0ae9968c3)

running 3 tests
test host::tests::execute ... ok
test item::tests::block ... ok
test tests::slice_len ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

rvolosatovs commented on src/v2/src/item/mod.rs:
```diff
@@ -18,8 +21,8 @@ impl TryFrom<usize> for Kind {
-            kind if kind == Kind::End as _ => return Ok(Kind::End),
-            kind if kind == Kind::Syscall as _ => return Ok(Kind::Syscall),
+            kind if kind == Kind::End as _ => Ok(Kind::End),
+            kind if kind == Kind::Syscall as _ => Ok(Kind::Syscall),
```
good point, I will fixup and force-push.

Commit:
v2: Add doc references to `item` types

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2/iter: Add a tail-recursive Iterator trait

`crate::iter::Iterator` allows implementing safe iterator over the
sallyport block.
`core::iter::Iterator::next` signature takes `&mut self` with anonymous
lifetime and therefore makes implementing a safe iterator splitting the
underlying mutable reference to a slice into two with distinct
lifetimes `'a` (tied to the lifetime of the block itself) impossible.

Implementation is an adaptation of `core::iter::Iterator` and provides
blanket implementations for `core::iter` traits. In other words, all
`core::iter::{,Into}Iterator` implementations implement `crate::iter::{,Into}Iterator`.

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2/{item, host}: Minimize use of `unsafe`

- Implement `crate::iter::Iterator`
- Refactor `item::Item` to allow for safe constructors of such type

Signed-off-by: Roman Volosatovs <roman@profian.com>

haraldh reviewed (commented):
LGTM, although I would turn `9` into some `SYSCALL_USIZE_NUM` const.

rvolosatovs commented:
> LGTM, although I would turn `9` into some `SYSCALL_USIZE_NUM` const.

Yes, absolutely, but the only thing I could come up with was `SYSCALL_USIZE_SIZE`, at which point I was not sure this would be an improvement anymore.
Maybe `SYSCALL_USIZE_COUNT`?

rvolosatovs commented:
Follow-ups for linting and documentation: #32 #31

haraldh commented:
Looks even better now

rvolosatovs commented:
Blocking #29
